### PR TITLE
Log receipt of deprecated or unsupported requests

### PR DESCRIPTION
**Pull request by gavin-norman-sociomantic**

### CODE REVIEW

**Comment by gavin-norman-sociomantic** [comment]
WIP implementation. Needs testing in a node. Blocked on having a neo node compatible with swarm v5.x.x.

**Comment by gavin-norman-sociomantic** [comment]
Tested (in dmqnode), updated, and unblocked.

**Comment by gavin-norman-sociomantic** [comment]
Updated with fixed stats log tests.

File: src/swarm/node/request/RequestStats.d [hunk 1]
```diff
@@ -745,4 +745,320 @@ public class RequestStats
...
+                    0 if rhs is equal to this instance, 1 otherwise
+
+            *******************************************************************/
+
+            mixin (genOpCmp("
+            {
+                if ( this.request == rhs.request && this.client == rhs.client )
+                    return 0;
+                else
+                    return 1;
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/node/request/RequestStats.d]: `return this.request != rhs.request || this.client != rhs.client` if you wish.

File: src/swarm/node/request/RequestStats.d [hunk 2]
```diff
@@ -745,4 +745,320 @@ public class RequestStats
...
+                this.client_request_info[ClientRequest(client.dup, rq.dup)]
+                    = ClientRequestInfo.init;
+                info = cl_rq in this.client_request_info;
+            }
+            assert(info !is null);
+
+            // Warn the first time a client sends an old request version.
+            if ( !info.active_this_hour(now) )
+                logger.warn("Client '{}' on {}:{} sent an old version of request '{}'",
+                    client, addr.address_bytes, addr.port, rq);
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/node/request/RequestStats.d]: I think the comment here should be made that this is triggered only once per hour per client per request, not only the first time (as far as I can see).

Comment (gavin-norman-sociomantic) [reply src/swarm/node/request/RequestStats.d]: Yes, that's correct :+1:

**Comment by gavin-norman-sociomantic** [comment]
Updated:
* Addressed Nemanja's points.
* Refactored the flag into `recent_activity`, which has a more intuitive function (as opposed to being a flag that just says "method X did Y").

File: src/swarm/neo/node/ConnectionHandler.d [hunk 3]
```diff
@@ -109,6 +109,12 @@ class ConnectionHandler : IConnectionHandler
             /// Indicates whether timing stats should be gathered about the
             /// request.
             bool timing;
+
+            /// Indicates that this request is scheduled for removal. The node
+            /// will handle it, but will log a warning.
+            // (Deliberately not named as a deprecation to avoid messing up
+            // searches for that keyword.)
+            bool scheduled_for_removal;
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/neo/node/ConnectionHandler.d:117]: Release notes missing.

File: src/swarm/node/request/RequestStats.d [hunk 4]
```diff
@@ -745,4 +745,327 @@ public class RequestStats
...
+            logger = Log.lookup("ScheduledForRemoval");
+        }
+
+        /// Tracks information about when a specific client sent a specific
+        /// scheduled-for-removal request.
+        private struct ClientRequestInfo
+        {
+            /// Indicates whether any recent activity has occurred for this
+            /// client/request pair. (We use this flag to essentially remove
+            /// items from the map (`client_request_info, below`), rather than
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/node/request/RequestStats.d]: The closing backtick is at the wrong position.

File: src/swarm/node/request/RequestStats.d [hunk 5]
```diff
@@ -745,4 +745,327 @@ public class RequestStats
...
+
+        public void started ( cstring rq, cstring client, AddrPort addr )
+        {
+            auto now = time(null);
+            auto cl_rq = ClientRequest(client, rq);
+            auto info = cl_rq in this.client_request_info;
+            if ( info is null )
+            {
+                this.client_request_info[ClientRequest(client.dup, rq.dup)]
+                    = ClientRequestInfo.init;
```
Comment (nemanja-boric-sociomantic) [inline src/swarm/node/request/RequestStats.d:1002]: I was worried that this opens an attack window, but the allocation happens only once per client per request :+1:.

File: relnotes/schedule-for-removal.feature.md [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+as long as the node maintainer decides it is needed.
+
+It is now possible to mark a request as "scheduled for removal". The request is
+handled as normal, but the node:
+1. Logs a warning.
+2. Outputs `old_request_versions_in_use:1` to its stats log.
+
+In this way, it is possible to tell when an older request version is no longer
+in use, rather than having to guess.
+
```
Comment (nemanja-boric-sociomantic) [inline relnotes/schedule-for-removal.feature.md:19]: Ah, here's the missing release note, but it should also say how the request is marked to be scheduled for removal.

**Comment by nemanja-boric-sociomantic** [comment]
Small notes about release notes and a small typo, but LGTM overall.

File: src/swarm/node/request/RequestStats.d (hunk 5 above)
Comment (gavin-norman-sociomantic) [reply src/swarm/node/request/RequestStats.d:1002]: Right.

**Comment by gavin-norman-sociomantic** [comment]
Updated and rebased.

**Review by nemanja-boric-sociomantic — APPROVED** [review]

**Pull request merged**